### ISSUE TITLE
Step4 Problem

### ISSUE THREAD

**Issue by NewtOliver** [unibas-gravis/Occlusion-Robust-MoFA#14]
Is it normal for me to have this situation in the last step of training the model?
![96aabe4cb1b9e5dae98c58510cf9880](https://user-images.githubusercontent.com/45253712/199507179-3a858671-3ef0-4537-b2c0-1ac458c78c7c.jpg)

**Comment by ChunLLee** [comment]
Hi, we do not expect such results. Can you check the following factors to provide more information?

1. Is the pre-trained model from step 1 working well?
2. Are the masks estimated by the UNet reasonable?
3. In the first 5000 iterations of step 4, are the faces rendered with normal human face texture?
4. If you run step 4 again, do such results occur?

**Comment by NewtOliver** [comment]
1. Step 1 is fine
2. I think the Pretrain_Unet is OK
3. In the first 5000 iterations of step 4, the network can normally display face texture, but the mask photo is all black, eg fig
4. Re-running step 4 still has the same problem, and the mask photo is all black from start to finish, Idon't know how to solve the problem
![image](https://user-images.githubusercontent.com/45253712/200262678-12818009-329f-4fd1-b323-2febba7f9be6.png)

**Comment by NewtOliver** [comment]
> Hi, we do not expect such results. Can you check the following factors to provide more information so that we could better diagnose?
> 
> 1. Is the pre-trained model from step 1 working well?
> 2. Are the masks estimated by the UNet reasonable?
> 3. In the first 5000 iterations of step 4, are the faces rendered with normal human face texture?
> 4. If you run step 4 again, do such results occur?

Step 1 is fine
I think the Pretrain_Unet is OK
In the first 5000 iterations of step 4, the network can normally display face texture, but the mask photo is all black, eg fig
Re-running step 4 still has the same problem, and the mask photo is all black from start to finish, Idon't know how to solve the problem
image

**Comment by NewtOliver** [comment]
> Hi, we do not expect such results. Can you check the following factors to provide more information so that we could better diagnose?
> 
> 1. Is the pre-trained model from step 1 working well?
> 2. Are the masks estimated by the UNet reasonable?
> 3. In the first 5000 iterations of step 4, are the faces rendered with normal human face texture?
> 4. If you run step 4 again, do such results occur?

1.Step 1 is fine
2.I think the Pretrain_Unet is OK
3.In the first 5000 iterations of step 4, the network can normally display face texture, but the mask photo is all black, eg fig
4.Re-running step 4 still has the same problem, and the mask photo is all black from start to finish, Idon't know how to solve the problem
![image](https://user-images.githubusercontent.com/45253712/201659764-1a6ec8fb-6ae5-43f2-98f2-d0b73ddecb92.png)

**Comment by hnuzhy** [comment]
I was also troubled by this problem or bug. Have you solved this problem during training Step4 ?

**Comment by NewtOliver** [comment]
> I was also troubled by this problem or bug. Have you solved this problem during training Step4 ?

Yes, I've solved. You should compare the code in the step4 with the code in the step1 and step3, and then fix it.

**Comment by DougDong666** [comment]
> 
I was also troubled by this problem. But the loss fuctions in step4 are different from step1 and step3, I don not think it can be fixed by changing loss functions in step3 and step1.

**Comment by DougDong666** [comment]
Loss functions in step4 are followed by article. But the loss function in step3 is a classic segmentation loss function. If the loss functions are changed by step3, I don not know why the article needs to mention its own loss fuctions.

**Comment by Jack171516** [comment]
Hi, can you take a look at the test_mask_gt.png image generated by your Step 3 Pretrain_Unet, I have a feeling that this image I generated might not be correct, which is causing me to train incorrectly in Step 3. Thank you very much.

**Comment by Jack171516** [comment]
Hi, can I have a look at your test_mask_gt.png image generated by Pretrain_Unet in step 3, I have a feeling that this image I generated might not be correct, which is causing me to train incorrectly in step 3. Thank you very much.

**Comment by JohnFive5** [comment]
@NewtOliver 
Hi, NewtOliver !
   I met the same problem when executing Step4_UNet_MoFA_EM.py. 

<img width="1314" height="860" alt="Image" src="https://github.com/user-attachments/assets/4640abde-8174-439d-8c61-07efd627ce43" />

<img width="1302" height="780" alt="Image" src="https://github.com/user-attachments/assets/fe8f6777-15de-4aeb-9004-785a330ac7e6" />

   The color of the generated textures are abnormal. And the generated masks are all dark. I made an ablation task on the Unet loss and finally found neighbor_unet_loss caused the problem. Can you give me some cues to fix the bug? Thanks a lot!
